app/scripts/controllers/permissions/index.js

```diff
@@ -0,0 +1,418 @@
+   *
+   * @param {string} origin - The origin string.
+   */
+  getAccounts (origin) {
+    return new Promise((resolve, _) => {
+
+      const req = { method: 'eth_accounts' }
+      const res = {}
+      this.permissions.providerMiddlewareFunction(
+        { origin }, req, res, () => {}, _end
+      )
+
+      function _end () {
+        if (res.error || !Array.isArray(res.result)) {
+          resolve([])
+        } else {
+          resolve(res.result)
+        }
+      }
+    })
+  }
+
+  /**
+   * Submits a permissions request to rpc-cap. Internal use only.
+   *
+   * @param {string} origin - The origin string.
+   * @param {IRequestedPermissions} permissions - The requested permissions.
+   */
+  _requestPermissions (origin, permissions) {
+    return new Promise((resolve, reject) => {
+
+      const req = { method: 'wallet_requestPermissions', params: [permissions] }
+      const res = {}
+      this.permissions.providerMiddlewareFunction(
+        { origin }, req, res, () => {}, _end
+      )
+
+      function _end (err) {
+        if (err || res.error) {
+          reject(err || res.error)
+        } else {
+          resolve(res.result)
+        }
+      }
+    })
+  }
+
+  /**
+   * User approval callback. The request can fail if the request is invalid.
+   *
+   * @param {object} approved the approved request object
+   */
+  async approvePermissionsRequest (approved, accounts) {
+
+    const { id } = approved.metadata
+    const approval = this.pendingApprovals[id]
+
+    try {
+
+      // attempt to finalize the request and resolve it
+      await this.finalizePermissionsRequest(approved.permissions, accounts)
+      approval.resolve(approved.permissions)
+
+    } catch (err) {
+
+      // if finalization fails, reject the request
+      approval.reject(ethErrors.rpc.invalidRequest({
+        message: err.message, data: err,
+      }))
+    }
+
+    delete this.pendingApprovals[id]
+  }
+
+  /**
+   * User rejection callback.
+   *
+   * @param {string} id the id of the rejected request
+   */
+  async rejectPermissionsRequest (id) {
+    const approval = this.pendingApprovals[id]
+    approval.reject(ethErrors.provider.userRejectedRequest())
+    delete this.pendingApprovals[id]
+  }
+
+  /**
+   * Grants the given origin the eth_accounts permission for the given account(s).
+   * This method should ONLY be called as a result of direct user action in the UI,
+   * with the intention of supporting legacy dapps that don't support EIP 1102.
+   *
+   * @param {string} origin - The origin to expose the account(s) to.
+   * @param {Array<string>} accounts - The account(s) to expose.
+   */
+  async legacyExposeAccounts (origin, accounts) {
+
+    const permissions = {
+      eth_accounts: {},
+    }
+
+    await this.finalizePermissionsRequest(permissions, accounts)
+
+    let error
+    try {
+      error = await new Promise((resolve) => {
+        this.permissions.grantNewPermissions(origin, permissions, {}, err => resolve(err))
```

Comment:
Nit: something like this would be a little more conventional:
```suggestion
        this.permissions.grantNewPermissions(origin, permissions, {}, err => err ? resolve() : reject(err))
```
Then the `error =` would be removed on the line above as well

Reply:
I made this change in 23c5e562c